internal/spicedbx/schema.go

```diff
@@ -8,26 +8,55 @@ import (
 	"go.infratographer.com/permissions-api/internal/types"
 )
 
-var (
-	schemaTemplate = template.Must(template.New("schema").Parse(`
+var schemaTemplate = template.Must(template.New("schema").Parse(`
+{{- define "renderCondition" -}}
+{{ $actionName := .Name }}
+{{- range $index, $cond := .Conditions -}}
+	{{- if $index }} + {{end}}
+	{{- if $cond.RelationshipAction }}
+		{{- $cond.RelationshipAction.Relation}}
+		{{- if ne $cond.RelationshipAction.ActionName ""}}->{{ $cond.RelationshipAction.ActionName }}{{- end }}
+	{{- end }}
+{{- end }}
+{{- end -}}
+
+{{- define "renderConditionSet" -}}
+{{ $actionName := .Name }}
+{{- range $index, $conditionSet := .ConditionSets }}
+	{{- if $index }} & {{end}}
+	{{- if gt (len $conditionSet.Conditions) 1 -}} ( {{- end}}
+	{{- range $index, $cond := .Conditions -}}
+		{{- if $index }} + {{end}}
+		{{- if $cond.RelationshipAction }}
+			{{- $cond.RelationshipAction.Relation}}
+			{{- if ne $cond.RelationshipAction.ActionName ""}}->{{ $cond.RelationshipAction.ActionName }}{{- end }}
+		{{- end }}
+	{{- end }}
+	{{- if gt (len $conditionSet.Conditions) 1 -}} ) {{- end}}
+	{{- end}}
+{{- end -}}
+
 {{- $namespace := .Namespace -}}
 {{- range .ResourceTypes -}}
 definition {{$namespace}}/{{.Name}} {
 {{- range .Relationships }}
-    relation {{.Relation}}: {{ range $index, $typeName := .Types -}}{{ if $index }} | {{end}}{{$namespace}}/{{$typeName}}{{- end }}
+    relation {{.Relation}}: {{ range $index, $type := .Types -}}
+			{{- if $index }} | {{end}}
+			{{- $namespace}}/{{$type.Name}}
+			{{- if $type.SubjectIdentifier}}:{{$type.SubjectIdentifier}}{{end}}
+			{{- if $type.SubjectRelation}}#{{$type.SubjectRelation}}{{end}}
+		{{- end }}
 {{- end }}
 
 {{- range .Actions }}
-    relation {{.Name}}_rel: {{ $namespace }}/role#subject
-{{- end }}
-
-{{- range .Actions }}
-{{- $actionName := .Name }}
-    permission {{ $actionName }} = {{ range $index, $cond := .Conditions -}}{{ if $index }} + {{end}}{{ if $cond.RoleBinding }}{{ $actionName }}_rel{{ end }}{{ if $cond.RelationshipAction }}{{ $cond.RelationshipAction.Relation}}->{{ $cond.RelationshipAction.ActionName }}{{ end }}{{- end }}
+    permission {{ .Name }} = {{ if gt (len .Conditions) 0 }}
+			{{- template "renderCondition" . }}
+		{{- else if gt (len .ConditionSets) 0 }}
+			{{- template "renderConditionSet" . }}
+		{{- end}}
```

Comment:
Defining one or the other seems unnecessary (and maybe confusing). Maybe we can converge on just one?

Reply:
Update: Looks like we can move that to a separate PR.